### ISSUE TITLE
Cargo stuck at "Updating crates.io"

### ISSUE THREAD

**Issue by rv178** [rust-lang/docker-rust#111]
I have been trying to get docker working for an app I was working on, but when I install packages, cargo gets stuck at "Updating crates.io" I tried some solutions, but none of them proved effective. Here's my Dockerfile:

```dockerfile
FROM alpine:3.16

WORKDIR /opt/app
COPY . .

RUN apk add --no-cache \
		ca-certificates \
		gcc \
		curl

ENV RUSTUP_HOME=/usr/local/rustup \
	CARGO_HOME=/usr/local/cargo \
	PATH=/usr/local/cargo/bin:$PATH \
	RUST_VERSION=1.62.1

RUN set -eux; \
	apkArch="$(apk --print-arch)"; \
	case "$apkArch" in \
		x86_64) rustArch='x86_64-unknown-linux-musl'; rustupSha256='bdf022eb7cba403d0285bb62cbc47211f610caec24589a72af70e1e900663be9' ;; \
		aarch64) rustArch='aarch64-unknown-linux-musl'; rustupSha256='89ce657fe41e83186f5a6cdca4e0fd40edab4fd41b0f9161ac6241d49fbdbbbe' ;; \
		*) echo >&2 "unsupported architecture: $apkArch"; exit 1 ;; \
	esac; \
	url="https://static.rust-lang.org/rustup/archive/1.24.3/${rustArch}/rustup-init"; \
	wget "$url"; \
	echo "${rustupSha256} *rustup-init" | sha256sum -c -; \
	chmod +x rustup-init; \
	./rustup-init -y --no-modify-path --profile minimal --default-toolchain $RUST_VERSION --default-host ${rustArch}; \
	rm rustup-init; \
	chmod -R a+w $RUSTUP_HOME $CARGO_HOME; \
	rustup --version; \
	cargo --version; \
	rustc --version;

RUN	rustup target add wasm32-unknown-unknown; \
	cargo install wasm-bindgen-cli;

RUN	curl -LO https://github.com/tailwindlabs/tailwindcss/releases/latest/download/tailwindcss-linux-x64; \
	chmod +x tailwindcss-linux-x64; \
	mv tailwindcss-linux-x64 /usr/local/bin/tailwindcss; \
	curl -L https://github.com/thedodd/trunk/releases/download/v0.16.0/trunk-x86_64-unknown-linux-gnu.tar.gz | tar -xzf-; \
	mv trunk /usr/local/bin/trunk; \
	cd frontend;

cmd ["trunk", "serve"]
```

**Comment by SET001** [comment]
had the same issue.
It stuck on 
```
  Installing api v0.1.0 (/usr/src/app)
    Updating crates.io index
```
for some time. And then throws errors

```
warning: spurious network error (2 tries remaining): [6] Couldn't resolve host name (Could not resolve host: static.crates.io)
warning: spurious network error (2 tries remaining): [6] Couldn't resolve host name (Could not resolve host: static.crates.io)
warning: spurious network error (1 tries remaining): [6] Couldn't resolve host name (Could not resolve host: static.crates.io)
warning: spurious network error (1 tries remaining): [6] Couldn't resolve host name (Could not resolve host: static.crates.io)
warning: spurious network error (1 tries remaining): [6] Couldn't resolve host name (Could not resolve host: static.crates.io)
error: failed to compile `api v0.1.0 (/usr/src/app)`, intermediate artifacts can be found at `/usr/src/app/target`

Caused by:
  failed to download from `https://crates.io/api/v1/crates/futures-core/0.3.21/download`

Caused by:
  [6] Couldn't resolve host name (Could not resolve host: static.crates.io)
```

Finally, I was managed to fix it by setting up DSN before run cargo install/build
```
RUN echo "nameserver 8.8.8.8" > /etc/resolv.conf && cargo install --path .
```

however, this looks like a temporary workaround and I think it should be a better solved in some other way.

**Comment by SET001** [comment]
Also, according to [this](https://docs.docker.com/engine/install/linux-postinstall/#specify-dns-servers-for-docker)  to solve this you can add this in your `/etc/docker/daemon.json`:
```
{
  "dns": ["8.8.8.8", "8.8.4.4"]
}
```

**Comment by roj1512** [comment]
I have the same problem.

**Comment by roj1512** [comment]
@SET001 your fixes aren't working.

**Comment by roj1512** [comment]
I'm on Arch Linux, too.

**Comment by rv178** [comment]
Yep, here too.

**Comment by Luchanso** [comment]
Hey, guys, the same problem, it can be reproduce with this docker file:
```Dockerfile
FROM rust:latest

WORKDIR /usr/src/myapp
COPY . .

RUN cargo install --path .

CMD ["myapp"]
```

On new project `cargo new --bin myapp` with any dependencies:
Cargo.toml:
```toml
[package]
name = "myapp"
version = "0.1.0"
edition = "2021"

[dependencies]
rand_core = "0.6.4"
```

without dependencies all working

**Comment by roj1512** [comment]
@Luchanso how much time did you give to it? I gave it like 30 minutes and then it worked.

**Comment by Luchanso** [comment]
@roj1512 about 15 minutes, but I'm not ready wait 30 minutes every time when I need new build
On my Mac (outside of docker) building take 1 - 2 min

I'll try later some older versions of base image for building (write here if they will work)

**Comment by roj1512** [comment]
@Luchanso The long wait was only once for me.

**Comment by Luchanso** [comment]
@roj1512 yes, because Docker have cache, but if you add new dependency, cache will invalidate and new build take again 30 min

**Comment by roj1512** [comment]
Exactly...

**Comment by Luchanso** [comment]
@roj1512 yes, I think so this happening because cargo trying download [large index](https://github.com/rust-lang/crates.io-index) (117 mb) very slow

<img width="838" alt="image" src="https://user-images.githubusercontent.com/2098777/192344301-b6d9d2b0-bdc9-4f3b-a83a-e58a7641e3c8.png">

# Workaround solution in single command
```Dockerfile
FROM rust:latest

# install crates.io index and cache it in the docker layer
RUN cargo new --bin build-index \
  && cd build-index \
  && cargo add rand_core \
  && cd .. \
  && rm -rf build-index

# ... other stuff ...
````
Explain:
1. Create empty fake project
2. Install any dependencies in this project
3. Cargo will build index
4. Remove fake project
5. Docker will cache this layer permanently, before you rewrite this command or adding some stuff before

Why is this a tradeoff? You need sometime reset cache layer, because index will outdate in time

# Workaround solution 2
[Or you can use rust nightly build and skip building index](https://stackoverflow.com/a/67001576), but I don't verified this way and don't know pitfalls

And I'll reopen issue https://github.com/rust-lang/cargo/issues/5721, because indicator not working in docker build

**Comment by Luchanso** [comment]
Interesting RFC about this problem https://rust-lang.github.io/rfcs/2789-sparse-index.html

**Comment by myssynglynx** [comment]
Having the same issue. I am able to add the dependencies using @Luchanso's workaround either in the Dockerfile or via terminal, but it throws the same error during the `cargo build --release` step. Does this have to do with internet connection? Or is it specific to particular processors? I am making sure the platform is adhering to the amd64 architecture, and I'm not on an M1 so it's not the [arm7 issue](https://github.com/rust-lang/docker-rust/issues/72).

Also, it works when I leave out certain dependencies in my `Cargo.toml`. For the server I'm trying to containerize, it builds without an issue if I leave out the `actix-web` & `sqlx` dependencies-- including either throws these `warning: spurious network error`s, despite building & running perfectly well on my local machine outside of docker.  

Note: I am using MacOs with a 2.3 GHz 8-Core Intel Core i9 processor, and 32 GB 2400 MHz DDR4 memory, and my internet connection is stable & fast (440 Mbps download, 20 Mpbs upload)

**Comment by Luchanso** [comment]
@myssynglynx no, internet connection not affected with it, this index (~117 mb) storing on github and github limiting download speed with 32 - 128 kb/sec, so that 117 mb / 32 kb/sec ≈ 60 min for downloading index.

Rust team can solve this problem with two ways:
1. Add sparse index
2. Buy hosting and run self-hosted index without speed limit, but this can be cost is really much

But I think any $6 hosting with unlimited traffic and bandwidth 200 mb/sec handle with this task

**Comment by Luchanso** [comment]
@myssynglynx I found second workaround solution, you can copy cargo cache index from your computer into docker container

For MacOS:
```sh
cd my-project
cp -r ~/.cargo/registry ./registry
```

For Linux: 
```sh
cd my-project
cp /usr/local/cargo/registry ./registry
```

Dockerfile:
```Dockerfile
FROM rust:alpine3.17

COPY registry /usr/local/cargo/registry
# .... other stuff ....
```

**Comment by myssynglynx** [comment]
@Luchanso This works. Thank you so much. It would be nice to not have to depend on anything from the developer's machine in order to get this running, but for now this will do.

**Comment by nouzu** [comment]
Using sparse-registry seems to solve the problem.

https://blog.rust-lang.org/2022/06/22/sparse-registry-testing.html
